### PR TITLE
Add EIP-5437: Interface for Security Contact

### PR DESCRIPTION
**Pull request by xinbenlv**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip-sc.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
+---
+eip: <to be assigned>
```
Comment (Pandapip1) [inline EIPS/eip-sc.md:2]:
```suggestion
eip: 5437
```

File: EIPS/eip-sc.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+requires: 165
+---
+
+An interface for smart contract developer to specify to communicate to them in secured way.
+A typical use case is for security researcher to report to developer the
+security vulnerability.
+
+## Abstract
+An interface for security notice using asymmetric encryption. The interface exposes a asymmetric encryption key
+and a destination of delivery.
```
Comment (SamWilsn) [inline EIPS/eip-sc.md:20]:
```suggestion
## Abstract

An interface for smart contracts to specify how to communicate with their developers using asymmetric encryption. For example, a security researcher may wish to report vulnerabilities over a secured channel. The interface exposes an asymmetric encryption key and a destination for delivery.
```

File: EIPS/eip-sc.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+## Motivation
+When security researchers identify security risk of a smart contract, currently there is no official way to
+report the problem to the smart contract maintainer, such as MakerDAO's issue.
+
+## Specification
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+```solidity
+interface ERC_Security_Contact {
```
Comment (SamWilsn) [inline EIPS/eip-sc.md:30]: Generally interfaces are named using `PascalCase` (so no underscores.)

File: EIPS/eip-sc.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+    /// REQUIRED
+    function getSecurityContact(bytes memory data) public view
+    returns (
+        uint8 type,
+        bytes memory publicKey,
+        bytes memory extraData
+    );
+
+    /// OPTIONAL
```
Comment (SamWilsn) [inline EIPS/eip-sc.md:40]: I'd recommend either splitting these optional functions into their own interface(s), or defining what they return when not implemented.

File: EIPS/eip-sc.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+report the problem to the smart contract maintainer, such as MakerDAO's issue.
+
+## Specification
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+```solidity
+interface ERC_Security_Contact {
+
+    /// REQUIRED
+    function getSecurityContact(bytes memory data) public view
```
Comment (SamWilsn) [inline EIPS/eip-sc.md:33]: Why does this take an argument?

I might also suggest the name `securityContact` instead.

File: EIPS/eip-sc.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        uint8 type,
+        bytes memory publicKey,
+        bytes memory extraData
+    );
+
+    /// OPTIONAL
+    function setSecurityContact(
+        uint8 type,
+        bytes memory publicKey,
+        bytes memory extraData) public;
```
Comment (SamWilsn) [inline EIPS/eip-sc.md:44]: I think you can omit setting the security contact from the interface entirely. It'll be project dependent anyway.

File: EIPS/eip-sc.md [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    event OnSetSecurityContact(uint8 type, bytes memory publicKeyForEncryption, bytes memory extraData);
```
Comment (SamWilsn) [inline EIPS/eip-sc.md:45]: I would suggest the name `SecurityContactChanged` instead.

File: EIPS/eip-sc.md [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+
+    /// OPTIONAL
+    function securityNotify(uint8 type, bytes memory data) public payable;
+    /// OPTIONAL
+    event OnSecurityNotification(uint8 type, bytes memory sourceData, uint256 value);
+
+    /// OPTIONAL
+    function bountyPolicy(uint id) public view returns(string);
```
Comment (SamWilsn) [inline EIPS/eip-sc.md:53]: I think this can be left out entirely, and possibly done as a separate EIP.

File: EIPS/eip-sc.md [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+1. Complying interface MUST implement the method of `getSecurityContact`.
+
+The `type` is a one byte data with valid range of `[0x10, 0x7f]`. The ranges of `[0x00, 0x0f]` and `[0x80, 0xff]` are preserved for future extension.
+
+The `type` indicates the format of the `publicKey` and `extraData` in the following way
+
+------------------------------------------------------------------------------------------------
+| Type | Encryption scheme                   | extraData                                       |
+-------|-------------------------------------|--------------------------------------------------
+| 0x10 | GnuPG - RSA/3072                    | Email address(es) encoded in format of RFC 2822 |
```
Comment (SamWilsn) [inline EIPS/eip-sc.md:66]: Would it also make sense to support a secp256k1 gpg algorithm?

File: EIPS/eip-sc.md (hunk 1 above)
Comment (xinbenlv) [reply EIPS/eip-sc.md]: Done! thank you!

File: EIPS/eip-sc.md (hunk 2 above)
Comment (xinbenlv) [reply EIPS/eip-sc.md]: Thank you, update according to your revision.

File: EIPS/eip-sc.md (hunk 3 above)
Comment (xinbenlv) [reply EIPS/eip-sc.md]: Updated to IEIP5437

File: EIPS/eip-sc.md (hunk 4 above)
Comment (xinbenlv) [reply EIPS/eip-sc.md]: Sounds good, would you be ok if I leave it to a follow up revision but to get this file check in first

File: EIPS/eip-sc.md (hunk 5 above)
Comment (xinbenlv) [reply EIPS/eip-sc.md]: I will provide detail in rationale for why there is an argument

File: EIPS/eip-sc.md (hunk 6 above)
Comment (xinbenlv) [reply EIPS/eip-sc.md]: That seems reasonable, I'd leave it here to triger design discussion for now. May remove when we decide not to have it. One argument would be regular rotation of public keys

File: EIPS/eip-sc.md (hunk 7 above)
Comment (xinbenlv) [reply EIPS/eip-sc.md]: Done, thanks for the suggestion

File: EIPS/eip-sc.md (hunk 8 above)
Comment (xinbenlv) [reply EIPS/eip-sc.md]: That sounds a good idea. Let me think about it. Added a TODo

File: EIPS/eip-sc.md (hunk 9 above)
Comment (xinbenlv) [reply EIPS/eip-sc.md]: I was actually thinking of it! Because RSA keys were so long and inefficient. Added a TODO

File: EIPS/eip-5437.md [hunk 10]
```diff
@@ -76,7 +76,9 @@ This table can be extended by future EIPs.
 in the table above.
 
 An example of `publicKey` using `RSA/3072` (TODO: decide whether to use secp256k1 instead)
-generated via [GnuPG](https://gnupg.org/),
+generated via GnuPG
+<!-- TODO waiting link to be allowed
+https://gnupg.org/ -->
```
Comment (Pandapip1) [inline EIPS/eip-5437.md:81]: This link will not be allowed

Comment (xinbenlv) [reply EIPS/eip-5437.md]: Updated to "add proper link to openstandard"

File: EIPS/eip-5437.md [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+------------------------------------------------------------------------------------------------
+
+This table can be extended by future EIPs.
+
+2. The `publicKey` returned from `getSecurityContact` MUST follow the encryption scheme specified
+in the table above.
+
+An example of `publicKey` using `RSA/3072` (TODO: decide whether to use secp256k1 instead)
+generated via GnuPG
+<!-- TODO add proper link to open standard for GPG -->
```
Comment (Pandapip1) [inline EIPS/eip-5437.md:80]:
```suggestion

```

No external links.

File: EIPS/eip-5437.md [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+a bounty policy.
+
+4. We intentionally add some additional information for future expandability.
+e.g. for `bountyPolicy` the `extraData` returned COULD potentially be used
+as the [EIP-5247](./eip-5247.md) bytecodes to specify a smart contract execution proposal.
+To stay less opinionated we leave this to the implementers. Future EIPs.
+are welcomed to extend specifying these.
+
+## Backwards Compatibility
+Currently, existing solutions such as OpenZeppelin use PlainText in source code
```
Comment (Pandapip1) [inline EIPS/eip-5437.md:162]:
```suggestion
Currently, existing solutions such as OpenZeppelin use plaintext in source code
```
PlainText™ is a registered trademark of the NonExistent corporation. Used without permission for laughs.

(/j obviously)

File: EIPS/eip-5437.md [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+```solidity
+/// @custom:security-contact some-user@some-domain.com
+```
+
+It's recommend that new version of smart contract to also adopt this EIP in-addition to
+this `@custom:security-contact` approach.
```
Comment (Pandapip1) [inline EIPS/eip-5437.md:169]:
```suggestion
It's recommend that new versions of smart contracts adopt this EIP in addition to the legacy `@custom:security-contact` approach.
```

File: EIPS/eip-5437.md [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+## Security Considerations
+
+Implementors should properly follow security practice required for the asymmetric schema
+to ensure a secured communication channel to be established when in need, such practices including but
+not limit too:
```
Comment (Pandapip1) [inline EIPS/eip-5437.md:175]:
```suggestion
Implementors should properly follow security practices required by the encryption scheme to ensure the security of the chosen communication channel. Some best practices are as follows:
```

File: EIPS/eip-5437.md [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+
+1. Keep security contact information up-to-date;
+2. Rotate encryption keys in the period recommended by best practice;
+3. Regularly monitor the incoming notification via the security contact information
+such as email inbox to ensure receiving the notice in a timely manner.
```
Comment (Pandapip1) [inline EIPS/eip-5437.md:180]:
```suggestion
3. Regularly monitor the channel to receive notices in a timely manner.
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
Please also run this through a spell/grammar checker. I've almost certainly missed a few typos or other errors.

**Comment by xinbenlv** [comment]
All suggestions committed. This commit mean to check in the file as Draft. I will run through grammar and spelling check prior to moving to Review status

**Comment by xinbenlv** [comment]
Request a manual merge @SamWilsn

**Comment by Pandapip1** [comment]
Why does this need a manual merge?

File: EIPS/eip-5437.md [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+
+The `type` indicates the format of the `publicKey` and `extraData` in the following way
+
+------------------------------------------------------------------------------------------------
+| Type | Encryption scheme                   | extraData                                       |
+-------|-------------------------------------|--------------------------------------------------
+| 0x10 | GnuPG - RSA/3072                    | Email address(es) encoded in format of RFC 2822 |
+------------------------------------------------------------------------------------------------
+
+This table can be extended by future EIPs.
```
[the remark this answers is not in the document]
Comment (Pandapip1) [reply EIPS/eip-5437.md:65]: My previous comment needs fixing.

File: EIPS/eip-5437.md [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+
+2. The `publicKey` returned from `getSecurityContact` MUST follow the encryption scheme specified
+in the table above.
+
+An example of `publicKey` using `RSA/3072` (TODO: decide whether to use secp256k1 instead)
+generated via GnuPG
+in a bytes which is ASCII(as in RFC 20) encoding of the public key string. Such as
```
[the remark this answers is not in the document]
Comment (Pandapip1) [reply EIPS/eip-5437.md:72]: My previous comment needs fixing here too.

File: EIPS/eip-5437.md [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+ndKG0bu7ZAi2b/c8uH/wHT5IvtfzHLeSTjDqG8UyLTnaDxHQZIE9JIzWSQ1DSoNC
+YrU7CQtL+/HRpiGFHfClaXln8VWkjnUvp+Fg1ZPtE1t/SKddZ7m29Hd9nzUc0OQW
+MOA+HDqgA3a9kWbQKSloORq4unft1eu/FCra
+=O6Bf
+-----END PGP PUBLIC KEY BLOCK-----
+```
+
+3. IF `setSecurityContact` is implemented and a call to it has succeeded in setting a new security
+contact, an event `SecurityContactChanged` MUST be emitted
+with the identical passed-in-parameters of `setSecurityContact`
```
[the remark this answers is not in the document]
Comment (Pandapip1) [reply EIPS/eip-5437.md:120]: My previous comment needs fixing here too.

File: EIPS/eip-5437.md [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+
+4. It's also RECOMMENDED that an on-chain security notify method `securityNotify`
+to implemented to receive security notice onchain. If it's implemented and a call
+has succeeded, it MUST emit an `OnSecurityNotification` with identical pass-in-parameter data.
+
+5. Compliant interface MUST implement [EIP-165](./eip-165.md).
```
[the remark this answers is not in the document]
Comment (Pandapip1) [reply EIPS/eip-5437.md:126]: My previous comment needs fixing here too.

File: EIPS/eip-5437.md [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+<!-- TODO: add EIP-165 interfaces. -->
+<!-- TODO also consider requiring/recommending implementing EIP-5629 ERC-interface detection. -->
+
+6. It's recommended to set a bounty policy via `bountyPolicy` method. The `id = 0` is preserved for a full overview, while other digits are used for different individual bounty policies. The returned
+string will be URI to content of bounty policies.
+No particular format of bounty policy is specified.
```
[the remark this answers is not in the document]
Comment (Pandapip1) [reply EIPS/eip-5437.md:132]: My previous comment needs fixing here too.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
Previous comments need fixing

File: EIPS/eip-5437.md [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+1. Complying interface MUST implement the method of `getSecurityContact`.
+
+The `type` is a one byte data with valid range of `[0x10, 0x7f]`. The ranges of `[0x00, 0x0f]` and `[0x80, 0xff]` are preserved for future extension.
+
+The `type` indicates the format of the `publicKey` and `extraData` in the following way
+
+------------------------------------------------------------------------------------------------
+| Type | Encryption scheme                   | extraData                                       |
+-------|-------------------------------------|--------------------------------------------------
+| 0x10 | GnuPG - RSA/3072                    | Email address(es) encoded in format of RFC 2822 |
```
[the remark this answers is not in the document]
Comment (xinbenlv) [reply EIPS/eip-5437.md:62]: Technical answer: While I appreciate your suggestion: yes we are still considering using EC-based algorithm but not determine yet.

I feel this is not editorial comment and shall not block merge of the file.

File: EIPS/eip-5437.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (xinbenlv) [reply EIPS/eip-5437.md:72]: I incline to keep it as is: 

secp256k1 is being considered but not rule out RSA/3072 for it's the default of OpenGPG. I don't think an technical but non-editorial comment shall block merging a DRAFT

File: EIPS/eip-5437.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (xinbenlv) [reply EIPS/eip-5437.md:65]: Let me reword

[the remark this answers is not in the document]
Comment (xinbenlv) [reply EIPS/eip-5437.md]: DONE

File: EIPS/eip-5437.md [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+-----END PGP PUBLIC KEY BLOCK-----
+```
+
+3. IF `setSecurityContact` is implemented and a call to it has succeeded in setting a new security
+contact, an event `SecurityContactChanged` MUST be emitted
+with the identical passed-in-parameters of `setSecurityContact`
+
+4. It's also RECOMMENDED that an on-chain security notify method `securityNotify`
+to implemented to receive security notice onchain. If it's implemented and a call
+has succeeded, it MUST emit an `OnSecurityNotification` with identical pass-in-parameter data.
```
[the remark this answers is not in the document]
Comment (xinbenlv) [reply EIPS/eip-5437.md:122]: I hold a different opinion in this. This is the last way of emergency notification or notification to DAO that doesn't have a centralized authority but still need to be notified.

File: EIPS/eip-5437.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (xinbenlv) [reply EIPS/eip-5437.md:130]: Thank you for this suggestion but I don't prefer this markdown style change. We don't have a particular flavor of markdown style.

File: EIPS/eip-5437.md [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+
+5. Compliant interface MUST implement [EIP-165](./eip-165.md).
+<!-- TODO: add EIP-165 interfaces. -->
+<!-- TODO also consider requiring/recommending implementing EIP-5629 ERC-interface detection. -->
```
[the remark this answers is not in the document]
Comment (xinbenlv) [reply EIPS/eip-5437.md:126]: Let's leave it a TODO

File: EIPS/eip-5437.md [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+MOA+HDqgA3a9kWbQKSloORq4unft1eu/FCra
+=O6Bf
+-----END PGP PUBLIC KEY BLOCK-----
+```
+
+3. IF `setSecurityContact` is implemented and a call to it has succeeded in setting a new security contact, an event `SecurityContactChanged` MUST be emitted with the identical passed-in-parameters of `setSecurityContact`
+
+4. It's also RECOMMENDED that an on-chain security notify method `securityNotify`
+to implemented to receive security notice onchain. If it's implemented and a call
+has succeeded, it MUST emit an `OnSecurityNotification` with identical pass-in-parameter data.
```
Comment (Pandapip1) [inline EIPS/eip-5437.md:122]:
```suggestion
4. It's also RECOMMENDED that an on-chain security notify method `securityNotify` to implemented to receive security notice onchain. If it's implemented and a call has succeeded, it MUST emit an `OnSecurityNotification` with identical pass-in-parameter data.
```

File: EIPS/eip-5437.md [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+
+5. Compliant interfaces MUST implement [EIP-165](./eip-165.md).
+<!-- TODO: add EIP-165 interfaces. -->
+<!-- TODO also consider requiring/recommending implementing EIP-5629 ERC-interface detection. -->
+
+6. It's recommended to set a bounty policy via `bountyPolicy` method. The `id = 0` is preserved for a full overview, while other digits are used for different individual bounty policies. The returned
+string will be URI to content of bounty policies.
+No particular format of bounty policy is specified.
```
Comment (Pandapip1) [inline EIPS/eip-5437.md:130]:
```suggestion
6. It's recommended to set a bounty policy via `bountyPolicy` method. The `id = 0` is preserved for a full overview, while other digits are used for different individual bounty policies. The returned string will be URI to content of bounty policies. No particular format of bounty policy is specified.
```

File: EIPS/eip-5437.md [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+------------------------------------------------------------------------------------------------
+| Type | Encryption scheme                   | extraData                                       |
+-------|-------------------------------------|--------------------------------------------------
+| 0x10 | GnuPG - RSA/3072                    | Email address(es) encoded in format of RFC 2822 |
+------------------------------------------------------------------------------------------------
+
+A new version of this table can be proposed by future EIPs by specifying a new `type` number.
+
+2. The `publicKey` returned from `getSecurityContact` MUST follow the encryption scheme specified
+in the table above.
```
Comment (Pandapip1) [inline EIPS/eip-5437.md:68]:
```suggestion
2. The `publicKey` returned from `getSecurityContact` MUST follow the encryption scheme specified in the table above.
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
There are still these formatting errors that need fixing.

These result in inconsistent rendering and are therefore *actually* blockers, not stylistic choices.

**Review by Pandapip1 — APPROVED** [review]
As much as I feel that there are issues with this, it's valid for a draft.

**Pull request merged**